### DISCUSSION TITLE
Help with temperature offsets

### DISCUSSION THREAD

**Discussion by mcharlee** [smartin015/continuousprint#189] [Q&A] [answered]
I'm not sure if maybe I have something set up wrong. I'm hoping someone out there can point me in the right direction. I have the temperature offsets assigned in the SpoolManager plugin but when I go to print through the ContinuousPrint plugin the offsets are not applied. Is there some setting that needs to be checked or updated when I am setting up the ContinuousPrint plugin? Is this a feature that is supported? Is there some workaround using a custom script? Thanks in advance!

**Comment by mcharlee** [comment]
Edit: Turns out unchecking and rechecking the temperature offsets did not solve the issue. However, as a workaround I found that if I select the spool before each item in the print job is printed the offsets get applied. Is it possible to have the offsets carry over or do I have to select the same spool every time I go to print something?

**Comment by smartin015** [comment]
Hey @mcharlee - I only use spoolmanager for filament tracking and haven't touched the temp offsets, so this is mostly exploratory.

It looks like the temp offset is applied [after the spoolmanager settings are saved](https://github.com/OllisGit/OctoPrint-SpoolManager/blob/dea8d64c1849c970f3616e158260c4c6fef5a4b7/octoprint_SpoolManager/__init__.py#L744), so that explains the beheavior you see (selecting the spool applies the offsets).

I can't find anywhere in OctoPrint's core code that would reset the offset after a print finishes, however. CPQ uses a pretty standard way for printing files as well, so I don't know why it would be ignoring the offset you apply. 

Can you confirm that the temperature offsets are applied on subsequent prints when printing *without* CPQ?

**Reply by mcharlee** [reply]
@smartin015 Thanks so much for the reply and for taking a look at that!

I use the temperature offsets so that I don't have to teach each user to update their slicer settings for each different type of filament that I have available for the printer. It really has come in handy that they can just set the default temperatures in their slicer, select the filament that they are using, and SpoolManager can apply the offsets after the fact. 

To answer, Yes I can confirm that the temperature offsets were applied to every print without CPQ. 

You'll have to forgive me cause I couldn't find where CPQ calls to SpoolManager, but I found [this line](https://github.com/OllisGit/OctoPrint-SpoolManager/blob/dea8d64c1849c970f3616e158260c4c6fef5a4b7/octoprint_SpoolManager/api/SpoolManagerAPI.py#L437) in the SpoolManagerAPI. Does this maybe give an idea?

**Reply by smartin015** [reply, marked as answer]
Thanks for confirming - that and your link actually tells me a great deal!

Looks like SpoolManager calls its api from the browser (see [here](https://github.com/OllisGit/OctoPrint-SpoolManager/blob/dea8d64c1849c970f3616e158260c4c6fef5a4b7/octoprint_SpoolManager/static/js/SpoolManager.js#L903)) when a print is about to start, in order to set the temperature offsets. Since CPQ sends its print command directly to OctoPrint, this flow is missed. 

It also clarifies #185 - since filament checks are done in JS as well, they're ignored by CPQ when starting a print. Looks like we'll need to integrate a bit more with SpoolManager in both cases by explicitly calling the `/allowedToPrint` and `/startPrintConfirmed` endpoints and handling the responses. 

I opened #191 to track this improvement. It'll take a little while to get this in - I'm away from home at the moment and there's a backlog of other changes I'm already committed to - but you can subscribe to that issue to get updates. Thanks again!
